Add subscription test case and Azure boot up case

yuxisun1217 opened the pull request:
Signed-off-by: Yuxin Sun <yuxisun@redhat.com>


This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] create a file in [news/unreleased](https://github.com/osbuild/osbuild-composer/tree/main/docs/news/unreleased) directory if this change should be mentioned in the release news
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

For user-visible changes, "adequate documentation" is an entry describing the
change for users in docs/news. Please refer to docs/news/README.md for details.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`
-->

atodorov reviewed (changes requested):
Code looks good but we need all the credentials properly configured in CI. ATM I see the following from `api.sh azure`

```ERROR: 
Missing credentials to access storage service.```

yuxisun1217 commented:
Hi @atodorov ,
I've update the script, removed the "dnf install dos2unix" step, because in some branches(e.g. https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1350313301 ) it cannot get available repo. Perhaps because this RHEL version has not been released yet and no published repo for that.
```
+ ssh -oStrictHostKeyChecking=no -i /tmp/tmp.iCaR9Uweoz/id_azure cloud-user@20.86.164.130 sudo dnf -y install dos2unix
Updating Subscription Management repositories.
Failed to set locale, defaulting to C.UTF-8
Error: There are no enabled repositories in "/etc/yum.repos.d", "/etc/yum/repos.d", "/etc/distro.repos.d".
```
I don't know now to check whether the OS is released and I think the package installation check is not so important. It'll be enough to just verify the system is subscribed. What's your opinion about that? Thanks!

atodorov commented on test/cases/api.sh:
```diff
@@ -89,6 +94,7 @@ case $CLOUD_PROVIDER in
     checkEnvAzure
     ;;
 esac
+checkEnvSubscription
```
Move to the case statement above and check only RHEL runners.

yuxisun1217 replied:
Thanks! Moved to RHEL only.

atodorov commented on test/cases/api.sh:
```diff
@@ -302,7 +325,14 @@ function createReqFileAWS() {
       "postgresql"
-    ]
+    ],
+    "subscription": {
```
subscription must be enabled only on RHEL runners I think. For Fedora and CentOS there's nothing to subscribe to.

atodorov replied:
I am surprised that nothing actually failed in this case. CC @Gundersanne

yuxisun1217 replied:
> > Hi @atodorov ,
> > I've update the script, removed the "dnf install dos2unix" step, because in some branches(e.g. https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1350313301 ) it cannot get available repo. Perhaps because this RHEL version has not been released yet and no published repo for that.
> > ```
> > + ssh -oStrictHostKeyChecking=no -i /tmp/tmp.iCaR9Uweoz/id_azure cloud-user@20.86.164.130 sudo dnf -y install dos2unix
> > Updating Subscription Management repositories.
> > Failed to set locale, defaulting to C.UTF-8
> > Error: There are no enabled repositories in "/etc/yum.repos.d", "/etc/yum/repos.d", "/etc/distro.repos.d".
> > ```
> 
> That's running on a rhel-8-x86_64 runner which is a released distro, 8.3 at this time. Something went wrong during subscription/repo enablement. Not sure what.

Hi @atodorov , do you think we still need this "package installation check" step? If not, we can just skip this issue. Thanks!

yuxisun1217 replied:
Thanks! Updated. I decoupled the subscription block from the whole json and make it optional. I know it's not good to read but I have no better idea...

atodorov commented:
> Hi @atodorov ,
> I've update the script, removed the "dnf install dos2unix" step, because in some branches(e.g. https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1350313301 ) it cannot get available repo. Perhaps because this RHEL version has not been released yet and no published repo for that.
> 
> ```
> + ssh -oStrictHostKeyChecking=no -i /tmp/tmp.iCaR9Uweoz/id_azure cloud-user@20.86.164.130 sudo dnf -y install dos2unix
> Updating Subscription Management repositories.
> Failed to set locale, defaulting to C.UTF-8
> Error: There are no enabled repositories in "/etc/yum.repos.d", "/etc/yum/repos.d", "/etc/distro.repos.d".
> ```
> 

That's running on a rhel-8-x86_64 runner which is a released distro, 8.3 at this time. Something went wrong during subscription/repo enablement. Not sure what.

yuxisun1217 commented:
> > Hi @atodorov ,
> > I've update the script, removed the "dnf install dos2unix" step, because in some branches(e.g. https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1350313301 ) it cannot get available repo. Perhaps because this RHEL version has not been released yet and no published repo for that.
> > ```
> > + ssh -oStrictHostKeyChecking=no -i /tmp/tmp.iCaR9Uweoz/id_azure cloud-user@20.86.164.130 sudo dnf -y install dos2unix
> > Updating Subscription Management repositories.
> > Failed to set locale, defaulting to C.UTF-8
> > Error: There are no enabled repositories in "/etc/yum.repos.d", "/etc/yum/repos.d", "/etc/distro.repos.d".
> > ```
> 
> That's running on a rhel-8-x86_64 runner which is a released distro, 8.3 at this time. Something went wrong during subscription/repo enablement. Not sure what.



> > Hi @atodorov ,
> > I've update the script, removed the "dnf install dos2unix" step, because in some branches(e.g. https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1350313301 ) it cannot get available repo. Perhaps because this RHEL version has not been released yet and no published repo for that.
> > ```
> > + ssh -oStrictHostKeyChecking=no -i /tmp/tmp.iCaR9Uweoz/id_azure cloud-user@20.86.164.130 sudo dnf -y install dos2unix
> > Updating Subscription Management repositories.
> > Failed to set locale, defaulting to C.UTF-8
> > Error: There are no enabled repositories in "/etc/yum.repos.d", "/etc/yum/repos.d", "/etc/distro.repos.d".
> > ```
> 
> That's running on a rhel-8-x86_64 runner which is a released distro, 8.3 at this time. Something went wrong during subscription/repo enablement. Not sure what.

atodorov commented:
> 
> Hi @atodorov , do you think we still need this "package installation check" step? If not, we can just skip this issue. Thanks!

The previous job that you linked failed because of the failing `dnf install` which reveals some issues with this job. Could be the runner, could be a network hiccup, IDK. The point is that without this failure we would never know about this. IMO the `dnf install` command should be present.

yuxisun1217 commented:
> > Hi @atodorov , do you think we still need this "package installation check" step? If not, we can just skip this issue. Thanks!
> 
> The previous job that you linked failed because of the failing `dnf install` which reveals some issues with this job. Could be the runner, could be a network hiccup, IDK. The point is that without this failure we would never know about this. IMO the `dnf install` command should be present.

Thanks Alex! Add it back. And try to auto-attach if no repo.
Besides, I also find some unexpected failures:
1. Sometimes the VM is not subscribed in 300s. See https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1351080822 and https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1351080820. 
2. Sometimes the "subscription-manager unregister" failed. See https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1351080825
```
+ ssh -oStrictHostKeyChecking=no -i ./keypair.pem cloud-user@3.143.244.165 sudo subscription-manager unregister
Unregistering from: subscription.rhsm.[MASKED]:443/subscription
Remote server error. Please check the connection details, or see /var/log/rhsm/rhsm.log for more information.
```
3. Sometimes the Azure VM fails to be created. Error details:
```
{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"OSProvisioningInternalError","message":"OS Provisioning failed for VM 'vm-b0b2d348-e5f6-4e00-b3df-881386dc6bec' due to an internal error: [ProtocolError] [CopyOvfEnv] Error mounting dvd: [OSUtilError] Failed to mount dvd device\nInner error: [mount -o ro -t udf,iso9660 /dev/sr0 /mnt/cdrom/secure] returned 32: mount: /mnt/cdrom/secure: no medium found on /dev/sr0.\n."}]}
```
It looks like an error from Azure side but I've never met it in my own testing. So I change a VM size to see if it happens again.

yuxisun1217 commented:
Hi @atodorov ,
Again I see there's no repo for RHEL-8.3: (https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1353875306)
```
+ ssh -oStrictHostKeyChecking=no -i ./keypair.pem cloud-user@3.135.228.119 sudo subscription-manager attach --auto
Installed Product Current Status:
Product Name: Red Hat Enterprise Linux for x86_64 Beta
Status:       Not Subscribed
Unable to find available subscriptions for all your installed products.
```
I guess the problem is that the system is identified as a "Beta". Not sure why it's beta. I've tried that created a RHEL-8.3 VM from Azure Marketplace and register with this RHSM account, it can get repo.

atodorov commented on test/cases/api.sh:
```diff
@@ -293,16 +316,32 @@ case $(set +x; . /etc/os-release; echo "$ID-$VERSION_ID") in
+      \"server-url\": \"subscription.rhsm.redhat.com\",
+      \"insights\": true
+    }"
+
```
Can we use a here document and avoid the escape symbols ?

yuxisun1217 replied:
Thanks! Good suggestion! Updated.

atodorov commented on test/cases/api.sh:
```diff
@@ -293,16 +316,32 @@ case $(set +x; . /etc/os-release; echo "$ID-$VERSION_ID") in
+    subscription="$SUBSCRIPTION_BLOCK"
+  else
+    subscription=''
+  fi
```
You can move this condition outside the function and conditionally define the value of `SUBSCRIPTION_BLOCK`, then avoid the repetition inside every function.

yuxisun1217 replied:
Yes you're right. Update.

atodorov reviewed (changes requested):
Small changes requested

atodorov commented:
>     1. Sometimes the VM is not subscribed in 300s. See https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1351080822 and https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1351080820.
> 

FAIL when this happens. Leave it as-is. If we see this happening often then we have to investigate.


>     2. Sometimes the "subscription-manager unregister" failed. See https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1351080825
> 
> 
> ```
> + ssh -oStrictHostKeyChecking=no -i ./keypair.pem cloud-user@3.143.244.165 sudo subscription-manager unregister
> Unregistering from: subscription.rhsm.[MASKED]:443/subscription
> Remote server error. Please check the connection details, or see /var/log/rhsm/rhsm.log for more information.
> ```
> 

>     1. Sometimes the Azure VM fails to be created. Error details:
> 
> 
> ```
> {"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"OSProvisioningInternalError","message":"OS Provisioning failed for VM 'vm-b0b2d348-e5f6-4e00-b3df-881386dc6bec' due to an internal error: [ProtocolError] [CopyOvfEnv] Error mounting dvd: [OSUtilError] Failed to mount dvd device\nInner error: [mount -o ro -t udf,iso9660 /dev/sr0 /mnt/cdrom/secure] returned 32: mount: /mnt/cdrom/secure: no medium found on /dev/sr0.\n."}]}
> ```
> 
> It looks like an error from Azure side but I've never met it in my own testing. So I change a VM size to see if it happens again.

atodorov commented:
> Hi @atodorov ,
> Again I see there's no repo for RHEL-8.3: (https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1353875306)
> 
> ```
> + ssh -oStrictHostKeyChecking=no -i ./keypair.pem cloud-user@3.135.228.119 sudo subscription-manager attach --auto
> Installed Product Current Status:
> Product Name: Red Hat Enterprise Linux for x86_64 Beta
> Status:       Not Subscribed
> Unable to find available subscriptions for all your installed products.
> ```
> 
> I guess the problem is that the system is identified as a "Beta". Not sure why it's beta. I've tried that created a RHEL-8.3 VM from Azure Marketplace and register with this RHSM account, it can get repo.

CC @ondrejbudai. Any idea what's happening here ?

ondrejbudai commented on test/cases/api.sh:
```diff
@@ -640,10 +730,33 @@ function verifyInAzure() {
+    # --ssh-key-values "$AZURE_SSH_KEY.pub" \
+    # --authentication-type "ssh" \
+### debug
+    --admin-password "$AZURE_VM_PASSWORD"
```
missing `\` at the end

yuxisun1217 replied:
Oh! Thanks!

yuxisun1217 commented:
New update: For Fedora images, skip boot up testing on Azure because of #1479

yuxisun1217 commented:
> New update: For Fedora images, skip boot up testing on Azure because of #1479

Reverted. Do not skip now.

yuxisun1217 commented:
> > Hi @atodorov ,
> > Again I see there's no repo for RHEL-8.3: (https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1353875306)
> > ```
> > + ssh -oStrictHostKeyChecking=no -i ./keypair.pem cloud-user@3.135.228.119 sudo subscription-manager attach --auto
> > Installed Product Current Status:
> > Product Name: Red Hat Enterprise Linux for x86_64 Beta
> > Status:       Not Subscribed
> > Unable to find available subscriptions for all your installed products.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I guess the problem is that the system is identified as a "Beta". Not sure why it's beta. I've tried that created a RHEL-8.3 VM from Azure Marketplace and register with this RHSM account, it can get repo.
> 
> CC @ondrejbudai. Any idea what's happening here ?

Hi @atodorov ,

I retried this job and checked the release version today. It is indeed RHEL-8.4 Beta. That's why it cannot find available subscriptions from CDN server. 
```
[root@vm-0955321c-3e71-4073-b569-9257d249b01f ~]# cat /etc/os-release 
NAME="Red Hat Enterprise Linux"
VERSION="8.4 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.4"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.4 Beta (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8.4:beta"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/red_hat_enterprise_linux/8/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.4
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.4 Beta"
```
The job is https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1360801578

atodorov commented:
So for the 8.4 runners (soon to be 8.5) - we can't subscribe to CDN b/c there's nothing to subscribe to really. The majority of the times these will be unreleased systems or at times (during the end of a release cycle) we'll be using older systems than what is available.

I'm not sure how to solve this problem ATM. IMO open a Jira ticket for the subscription+nightly part and we can deal with it later. Let's now focus on having all jobs for this PR report PASS so we can merge it.

yuxisun1217 commented:
> So for the 8.4 runners (soon to be 8.5) - we can't subscribe to CDN b/c there's nothing to subscribe to really. The majority of the times these will be unreleased systems or at times (during the end of a release cycle) we'll be using older systems than what is available.
> 
> I'm not sure how to solve this problem ATM. IMO open a Jira ticket for the subscription+nightly part and we can deal with it later. Let's now focus on having all jobs for this PR report PASS so we can merge it.

Opened a ticket for that: https://issues.redhat.com/browse/COMPOSER-996
And I just updated the case to skip package installation checking in Beta versions.

atodorov reviewed (changes requested):
Rebase to matest `main`. Testing on Fedora32 has been removed.

yuxisun1217 commented:
> Rebase to matest `main`. Testing on Fedora32 has been removed.

Rebased and changed the resource delete order in Azure cleanup due to https://github.com/osbuild/osbuild-composer/issues/1479#issuecomment-865659613

ondrejbudai commented:
So the rpmrepo snapshot that we're using for Fedora 33 contains `WALinuxAgent-2.2.48.1`. If we want a newer one, we need to update the snapshot in `test/data/repositories`. You can find how to find the repository baseurl here: https://www.osbuild.org/rpmrepo/

yuxisun1217 commented:
Blocked by #1487 .

yuxisun1217 commented:
> > So for the 8.4 runners (soon to be 8.5) - we can't subscribe to CDN b/c there's nothing to subscribe to really. The majority of the times these will be unreleased systems or at times (during the end of a release cycle) we'll be using older systems than what is available.
> > I'm not sure how to solve this problem ATM. IMO open a Jira ticket for the subscription+nightly part and we can deal with it later. Let's now focus on having all jobs for this PR report PASS so we can merge it.
> 
> Opened a ticket for that: https://issues.redhat.com/browse/COMPOSER-996
> And I just updated the case to skip package installation checking in Beta versions.

Revert the skip because the Beta SKU has been added into the test account. Just need to disable all *beta* repos because they are unreachable. Code has been updated.

atodorov commented on test/cases/api.sh:
```diff
@@ -613,6 +665,46 @@ function _instanceWaitSSH() {
+    $_ssh sudo rm -rf /var/cache/dnf
+    # Verify package installation. Disable *beta* repos because they are not reachable
+    $_ssh sudo dnf repolist
+    $_ssh sudo dnf -y install dos2unix --disablerepo=*beta*
```
hmmm. I'm wondering now that you have the beta SKUs on the account isn't this still going to cause dnf failure if the beta repos are disabled?

yuxisun1217 replied:
So I still skip the subscription check in Beta images now...I disable *beta* repos here just in case they blocks the non-beta images. I'm not quite sure if they will appear in GA images.

yuxisun1217 replied:
I'm asking RHSM members if they know why the beta repos are unreachable. I still keep the https://issues.redhat.com/browse/COMPOSER-996 open and I'll update there if I get the answer.

atodorov replied:
Asking here in context again. 

If your code is skipping subscriptions on a Beta system (plus that you have a Beta SKU) what's the current problem then ? From all I can tell this section of the script should be skipped so you shouldn't worry about available repos when running on the 8.4 runners?

yuxisun1217 replied:
I think you're right. There shouldn't be beta repos in non-beta systems. I'll remove the "--disablerepo=*beta*" part. Thanks!

ondrejbudai commented:
Do we even want to run this test on pre-GA RHEL? Image Builder runs on GA RHEL and produces only GA RHEL images. There's definitely some value in testing pre-GA stuff but if it's too weird with subscriptions, I would suggest just dropping it. AFAIK beta repositories are removed shortly after GA.

yuxisun1217 commented:
> Do we even want to run this test on pre-GA RHEL? Image Builder runs on GA RHEL and produces only GA RHEL images. There's definitely some value in testing pre-GA stuff but if it's too weird with subscriptions, I would suggest just dropping it. AFAIK beta repositories are removed shortly after GA.

Hi Ondrej,

Actually I'm not quite understand what is the "rhel-8.json" for? Is it for the latest nightly RHEL-8 compose test? Thanks!

ondrejbudai commented:
Do you mean `test/data/repositories/rhel-8.json`? These are repositories are used e. g. for test manifests generation, api tests, koji tests, `grep -r /usr/share/tests/osbuild-composer/repositories .` might help you.

atodorov commented:
> Do we even want to run this test on pre-GA RHEL? Image Builder runs on GA RHEL and produces only GA RHEL images. There's definitely some value in testing pre-GA stuff but if it's too weird with subscriptions, I would suggest just dropping it. AFAIK beta repositories are removed shortly after GA.

Probably not much, let's discuss this on the QE meeting in few days. 

OTOH a Beta system already skips the subscription part, that's:
```
  # Beta image doesn't have available subscription so skip subscription check. See COMPOSER-996
  if [[ "$ID" =~ "rhel" ]] && [[ ! "$pretty_name" =~ 'Beta' ]]; then
```

so what's the current issue we're trying to resolve then ?

atodorov commented on test/cases/api.sh:
```diff
@@ -326,6 +352,23 @@ case $(set +x; . /etc/os-release; echo "$ID-$VERSION_ID") in
 esac
 
+# Only RHEL need subscription block.
+if [[ "$ID" == "rhel" ]]; then
```
add ` && [ ! "$pretty_name" =~ 'Beta' ]`

then for Beta images SUBSCRIPTION_BLOCK is empty.

yuxisun1217 replied:
About this part I'm also a little confused. It seems that the base machine and the guest VM are not the same RHEL versoin. See https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1376400878 , at the beginning, the /etc/os-release shows the system is RHEL-8.3(https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1376400878#L392):
```
$ schutzbot/deploy.sh
+ PROJECT=osbuild-composer
+ source /etc/os-release
++ NAME='Red Hat Enterprise Linux'
++ VERSION='8.3 (Ootpa)'
++ ID=rhel
```
But when I get the pretty_name from the guest VM it shows the system is RHEL-8.4 Beta:(https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1376400878#L3390)
```
++ ssh -oStrictHostKeyChecking=no -i /tmp/tmp.f6VSJ5cul0/id_azure cloud-user@104.40.137.221 grep '^PRETTY_NAME=' /etc/os-release
+ pretty_name='PRETTY_NAME="Red Hat Enterprise Linux 8.4 Beta (Ootpa)"'
```

So that I cannot get the pretty_name before the guest VM is actually boot up.

atodorov replied:
I'm not quite sure what happens here either but I'm sure it has to do with the fact that the content for the images is decoupled from the host operating system. It's probably coming from the repository json files shipped together with the RPM. 

But that goes to show another thing - you can't rely on the host OS to make these decisions. The decision should be based on some external variable which we're going to pass down to this test script in the test matrix.  Similar to how we pass INTERNAL_NETWORK.

This is becoming a mess quickly. 

Option 1) Split out the Azure boot up part of the code in a separate PR - that should not have any issues anymore and we can merge it as soon as CI reports pass. Leave the subscription portion here (after a rebase).

Option 2) Wait until we discuss and make a decision whether or not there's value to persue subscription testing on unreleased targets and then make a decision how to implement it to avoid all of these problems.

yuxisun1217 replied:
Hi @atodorov ,
Thanks for your suggestion! I think use external variable is a good idea. But I don't wish to split this PR because there are already too many discussions/changes in this PR and I don't want to make it too messy. IMO the current solution(get the OS info from the guest) is not "wrong" but only not graceful enough. Shell we consider to accept the current solution and open a new issue to change it to the external variable way later? Thanks again!

atodorov replied:
Few comments here: 

* I am proposing the split only of the boot portion of this test which doesn't appear to have issues and the majority of the discussion is not about that. So there's not much to lose in practicality here.  This is if you don't want the second part of this PR to hold-back a portion of this test script which appears to be functioning fine.

* I don't see the point of merging a sub-optimal implementation and then following with more changes when:  1) we've identified a clear problem and possible solution 2) the current PR is still open and 3) we are not in a hurry to merge the current PR immediately. We have spent a significant time here already and it doesn't really make any difference if we merge now or spend some more time to adjust the implementation given all of the new information discovered in the last few days.  1 or 2 PRs doesn't matter here, in any case the test script will be WIP.

atodorov replied:
@jrusz, @ondrejbudai what are your preferences here ^^^ ?

yuxisun1217 replied:
Thank you for your explanation @atodorov ! I agree that we can discuss more about the new external variable in the meeting. I've opened an issue(#1508)

atodorov commented on test/cases/api.sh:
```diff
@@ -613,6 +665,46 @@ function _instanceWaitSSH() {
+
+  pretty_name=$($_ssh grep '^PRETTY_NAME=' /etc/os-release)
+  # Beta image doesn't have available subscription so skip subscription check. See COMPOSER-996
+  if [[ "$ID" =~ "rhel" ]] && [[ ! "$pretty_name" =~ 'Beta' ]]; then
```
see my comment above and check only if SUBSCRIPTION_BLOCK isn't empty. Then we'll have the selection logic in one place above and this function doesn't need to know on what OS variant/version it is running.

atodorov commented on test/cases/api.sh:
```diff
@@ -613,6 +665,46 @@ function _instanceWaitSSH() {
+    # Clean cache and get enabled repo list
+    $_ssh sudo rm -rf /var/cache/dnf
+    # Verify package installation. Disable *beta* repos because they are not reachable
+    $_ssh sudo dnf repolist
```
B/c the origins of this code-block were copied from the image-builder repository and b/c the `rpm -q dos2unix` check was removed in https://github.com/osbuild/image-builder/pull/190#discussion_r644010182 and b/c this PR is still open IMO it makes sense to just add the check back. This is equivalent to fixing https://github.com/osbuild/image-builder/issues/195 inside the osbuild-composer repository.

yuxisun1217 replied:
Hi Alex,
I've removed all the other repos before installation. And due to https://github.com/osbuild/image-builder/pull/190#discussion_r644002358 we have confirmed that if there's no available repo, it will returns 1 no matter whether this package is installed or not. So that I think there's no need to add the pre-check anymore. Is that acceptable? Thanks!

atodorov replied:
Show me where the repos are removed b/c I am not seeing it.

atodorov replied:
That's a hell of a weird way to make sure your pre-condition is met. Let's hope dnf doesn't change behavior.

yuxisun1217 commented:
> Do you mean `test/data/repositories/rhel-8.json`? These are repositories are used e. g. for test manifests generation, api tests, koji tests, `grep -r /usr/share/tests/osbuild-composer/repositories .` might help you.

Thanks Ondrey! I meant to ask that what are the differences between "rhel-8" and "rhel-8.4" branches? It seems that the repos in rhel-8.4.json point to the latest 8.4 development composes. But I've no idea what is "https://rpmrepo.osbuild.org/v1/psi/el8/el8-x86_64-baseos-8.4.0.n-20201108" in rhel-8.json and why we test it. Thanks!

yuxisun1217 commented:
Updated the PR due to the discussion today:
Do not verify package installation in api.sh. Only verify subscription registration in all the images. 
There's only 1 concern:
Currently we use the host OS version(ID) to judge if we need to test the subscription registration case in guest image. Due the the previous discussion we know that the host OS and guest OS are not the same. So it'll be better to add an external variable here. But currently it's not a problem because the product ID(rhel/fedora) of the host and guest OS are the same.

atodorov commented:
> Updated the PR due to the discussion today:
> Do not verify package installation in api.sh. Only verify subscription registration in all the images.
> There's only 1 concern:
> Currently we use the host OS version(ID) to judge if we need to test the subscription registration case in guest image. Due the the previous discussion we know that the host OS and guest OS are not the same. So it'll be better to add an external variable here. But currently it's not a problem because the product ID(rhel/fedora) of the host and guest OS are the same.

This is not going to be a practical problem until the time when building cross distro images becomes a supported functionality.

atodorov reviewed (changes requested):
`sudo: subscription-manager: command not found`

Add subscription-manager to the dependencies in spec file.

ondrejbudai commented:
This appeared on Fedora, no? I'm not sure whether we want to run this test on Fedora. I think it doesn't bring much value.

yuxisun1217 commented:
> `sudo: subscription-manager: command not found`
> 
> Add subscription-manager to the dependencies in spec file.

Hi Alex and Ondrej,

So sorry I made a stupid mistake while rebasing. I'd like to debug the CI results myself first and then when it's all passed I'll @you to help review it, to prevent wasting your time :)

atodorov commented:
> This appeared on Fedora, no? I'm not sure whether we want to run this test on Fedora. I think it doesn't bring much value.

That's a side effect of how the matrix is defined.

yuxisun1217 commented:
Hi @atodorov , all the jobs are passed. Could you please help to merge it when available? Thank you so much!

atodorov commented on test/cases/api.sh:
```diff
@@ -467,6 +510,8 @@ function createReqFileAzure() {
 }
 EOF
+### debug
+  cat "$REQUEST_FILE"
```
Left over debug line.  If you want to keep it then use echo or green_print to add some delimiters in the output so this can be easily found for future reference. Alternaively you may want to upload this as an artifact in GitLab CI. 

Note: when printing on the terminal GitLab seems to do a good job of masking secret credentials, not sure what will happen if we upload the file as artifact.

yuxisun1217 replied:
Thanks @atodorov ! I also hesitate if I should left this debug info. But for safe I'd like to remove it. 
Only changed this part in this push:
-### debug
-  cat "$REQUEST_FILE"

ondrejbudai reviewed (commented):
Looks good.

atodorov reviewed (approved):
LGTM now. Waiting for CI jobs to report PASS.

atodorov commented:
Azure test jobs reported PASS. Merging.
